examples/react-navigation/README.md

```diff
@@ -0,0 +1,9 @@
+# RNTL example app for React Navigation
+
+This example shows how to write integration tests using React Navigation without mocking it.
+
+There are two types of tests:
+1. integration tests operating on whole navigators, they should use `renderNavigator` helper to render a navigator component used in the app. It is useful when you want to test a scenario that includes multiple screens.
+2. single screen tests where you would pass mock `navigation` prop, built using `buildNavigationMock()` helper, and `route` prop to the screen component using regular `render` function.
```

Comment:
I wouldn't show this approach since this will only work for very simple cases. For more involved use cases, mocking the navigation object can quickly get complex ad the available methods are different based on the navigator and nesting. There are also other hooks that would fail when not used in a navigator.

Instead, I'd suggest writing components that don't use anything React Navigation specific and instead accept explicit callbacks and props for what they need, so that it's easy to test them in isolation.

Reply:
@satya164 to understand you correctly, instead of mocking React Nav props/hooks for single screen test, you would rather refactor the code of e.g. `DetailsScreen` (that accepts `route` & `navigation` props) by
* extracting `DetailsScreenContent` that accepts callbacks like `onNavigateBack`, `onNavigateToSettings(...)`, etc
* make `DetailsScreen` a thin wrapper over `DetailsScreenContent` so that it only provides the above callbacks and connects them to navigation actions.

Is that correct or would you do it in some other way?

If that is the case, I think it would be painful for using wanting to test apps using React Nav hooks, since it will require them to refactor each of these components into navigation-less `Content` and navigation `Wrapper` components.

Reply:
I think testing using navigators is not always feasible, e.g. in case you want to test various scenarios that are on a screen "late" in the stack navigator. I think that in this case user would still be content being able to verify that `navigation.navigate(...)` or other relevant methods were/weren't called with relevant route and params.

@satya164 Would it be possible/feasible to have some mock `TestNavigator` that could that could be used to verify assertion. It could be used in following setup:

```jsx
const Test = createTestNavigator();

test("...", () => {
  render(
    <NavigationContainer>
      <Test.Navigator>
        <Test.Screen component={ScreenUnderTest} initialParams={{ ... }} />
      </Test.Navigator>
    <NavigationContainer>
  );

  fireEvent.press(getByText('Some button'));  
  expect(Test.navigation.navigate).toHaveBeenCalledWith(['Details', { id: 5 }]);
}
```

Such mock would be a full navigator but it would render only a single screen and expose `navigation` object so that user would be able to verify if they were called with proper params.

Reply:
There's a bit of a disconnect here, for me, there are 2 areas for testing:

1. Testing screens: Screens can often use navigator-specific methods, rely on navigation lifecycle, navigator-specific behavior, etc. There's no good way to test a screen properly without rendering the navigator. You can mock some stuff, but different screens will need different stuff mocked and in the end, you end up replicating React Navigation's API but it may not have the same behavior. These are integration tests - so we need to test them in an appropriate environment that doesn't behave too differently.

And we're only talking about navigation-related stuff. A screen would usually contain data fetching and other things as well which will need to be mocked.

If we're testing whether `navigation.navigate` is called or not, it's not the right thing to test for. If we want to test that user goes to `Details` screen, then we should check if we're on that screen. The `navigation.navigate` method is an implementation detail - if someone changes its signature or uses a different method like `navigation.jumpTo`, it may do the exact same thing but will break the tests.

2. Testing components: These are unit tests for components that might use React Navigation's API. Mocking is less problematic here, but it's still difficult. One approach would be to write a test navigator (react-navigation can't provide one since all navigators can behave differently and can have different methods). Due to this, I would really recommend dependency injection - pass the callbacks it needs. It'd simplify the tests a lot and ensure that we're only testing the public API of the component, i.e. what the component exposes.

> I think that in this case user would still be content being able to verify that navigation.navigate(...) or other relevant methods were/weren't called with relevant route and params.

People may be content with doing them, but imo it's important to suggest best practices and discourage testing implementation details in official docs. Testing the actual thing is harder, yes, but it also ensures that the tests actually help identify issues and don't fail due to a refactor. I am pretty sure people would still mock and test implementation details, but at least they'd have an idea of what a good test should look like by reading official docs. After all, these are just guidelines, people don't have to do it the exact same way - but they'll get more value from their tests if they do.

I'd also bring up the guiding principles of testing library which are in-line with how I see this: [The more your tests resemble the way your software is used, the more confidence they can give you.](https://testing-library.com/docs/guiding-principles) and Avoid testing [implementation details](https://kentcdodds.com/blog/testing-implementation-details).